# 01 Data Description












You can attempt any one:

A Data pertaining  to job seekers in India for the year (2021). Four files are named as per the number of records they contain while the fifth file is named as test.xls.
The variables are self explanatory. However, the entries will need some cleaning up. 


Special notes: Some of the salary figures are in Rs while some others are in different currencies. Consider only those which are in Rs
Lac indicates 100000
If you want to incorporate any other data (at the city level for control, you can do that. Mention your data source)

Question: Ask three interesting questions from the data and answer them. Question 1 must be solved using Clustering; Question 2 using Linear Regression and Question 3 using Logistic Regression.

Marks:
Data preparation and cleaning: 10 marks
Q1 (Clustering): 10 marks
Q2( Linear Regression): 10 marks
Q3 (Logistic Regression): 10 marks
Overall recommendation and findings based on the model: 10 marks

In [5]:
import pandas as pd

In [ ]:
df = pd.read_excel("../Job Seekers.xlsx")
df.head()

City                  DOB  Gender  \
0   HARIDWAR  1986-11-08 00:00:00     NaN   
1  AHMEDABAD  1989-05-30 00:00:00     NaN   
2      ANAND  1979-12-03 00:00:00     NaN   
3   AZAMGARH  1983-08-02 00:00:00     NaN   
4   ANANTNAG  1981-03-23 00:00:00     NaN   

                               Functional_Area Area_of_Specialization  \
0       Fresh Graduate-No Functional Area exp.       Other Fresh Grad   
1       Fresh Graduate-No Functional Area exp.       Other Fresh Grad   
2  Ticketing/Reservation/Travel Agent/Airlines            Air Hostess   
3                                      Finance     Financial Analysis   
4                  Marketing/MR/Media Planning         Media Planning   

                                    Industry  \
0                                  Retailing   
1                                     Others   
2  Hotel/Travel/Tourism/Airlines/Hospitality   
3         Accounting-Tax/Consulting, Banking   
4                        Entertainment/Media   

                                        Resume_Title  \
0  Pursuing B.Com last year, and with 6 mon diplo...   
1  Fresh Graduate looking for Job in a co-orporat...   
2                                              priti   
3                                      ashish mishra   
4                   marketing and sales related jobs   

                                          Key_Skills         Salary  \
0  MS Office,DTP, Internet and smooth operation o...  Rs. 4.88 lacs   
1  BOARD/UNIVERSITY YEAR OF PASSING B.A Manipur U...  Rs. 12.2 lacs   
2  diploma in aviation,travel and hospitality fro...  Rs. 7.89 lacs   
3                                            FRESHER   Rs. 7.0 lacs   
4                                institutional sales   Rs. 6.8 lacs   

        Work_Experience                          Level Preferred_Location  \
0  6 Year(s) 3 Month(s)            Entry Level/Fresher              Delhi   
1  8 Year(s) 2 Month(s)            Entry Level/Fresher              Delhi   
2           22 Year(s)             Entry Level/Fresher              Delhi   
3            9 Year(s)             Entry Level/Fresher  Anywhere in India   
4  9 Year(s) 5 Month(s)  Junior - Executive, Assistant         Delhi/ NCR   

  Course(Highest_Education) Specialization(Highest_Education)  \
0                     B.Com                   Commerce Honors   
1                      B.A.                       Pass Course   
2                      12th                           Science   
3                     B.Com                   Commerce Honors   
4                  MBA/PGDM                         Marketing   

  Institute(Highest_Education) Course(2nd_Highest_Education)  \
0            Jiwaji University                           NaN   
1                       Others                           NaN   
2                       Others                          B.A.   
3                       Others                         B.Com   
4    Aligarh Muslim University                         B.Com   

  Specialization(2nd_Highest_Education) Institute(2nd_Highest_Education)  
0                                   NaN                              NaN  
1                                   NaN                              NaN  
2                          Anthropology                           Others  
3                       Commerce Honors                           Others  
4                         Commerce Pass                 Delhi University

In [7]:
df.shape

(913384, 18)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 913384 entries, 0 to 913383
Data columns (total 18 columns):
 #   Column                                 Non-Null Count   Dtype  
---  ------                                 --------------   -----  
 0   City                                   236430 non-null  object 
 1   DOB                                    913384 non-null  object 
 2   Gender                                 0 non-null       float64
 3   Functional_Area                        908820 non-null  object 
 4   Area_of_Specialization                 826919 non-null  object 
 5   Industry                               876871 non-null  object 
 6   Resume_Title                           910254 non-null  object 
 7   Key_Skills                             826915 non-null  object 
 8   Salary                                 833386 non-null  object 
 9   Work_Experience                        904024 non-null  object 
 10  Level                                  897760 non-null  

In [13]:
df.describe()

Gender
count     0.0
mean      NaN
std       NaN
min       NaN
25%       NaN
50%       NaN
75%       NaN
max       NaN

In [17]:
# Handle invalid dates by using errors='coerce' to convert problematic dates to NaT
df['DOB'] = pd.to_datetime(df['DOB'], errors='coerce')
df.head()

City        DOB  Gender                              Functional_Area  \
0   HARIDWAR 1986-11-08     NaN       Fresh Graduate-No Functional Area exp.   
1  AHMEDABAD 1989-05-30     NaN       Fresh Graduate-No Functional Area exp.   
2      ANAND 1979-12-03     NaN  Ticketing/Reservation/Travel Agent/Airlines   
3   AZAMGARH 1983-08-02     NaN                                      Finance   
4   ANANTNAG 1981-03-23     NaN                  Marketing/MR/Media Planning   

  Area_of_Specialization                                   Industry  \
0       Other Fresh Grad                                  Retailing   
1       Other Fresh Grad                                     Others   
2            Air Hostess  Hotel/Travel/Tourism/Airlines/Hospitality   
3     Financial Analysis         Accounting-Tax/Consulting, Banking   
4         Media Planning                        Entertainment/Media   

                                        Resume_Title  \
0  Pursuing B.Com last year, and with 6 mon diplo...   
1  Fresh Graduate looking for Job in a co-orporat...   
2                                              priti   
3                                      ashish mishra   
4                   marketing and sales related jobs   

                                          Key_Skills         Salary  \
0  MS Office,DTP, Internet and smooth operation o...  Rs. 4.88 lacs   
1  BOARD/UNIVERSITY YEAR OF PASSING B.A Manipur U...  Rs. 12.2 lacs   
2  diploma in aviation,travel and hospitality fro...  Rs. 7.89 lacs   
3                                            FRESHER   Rs. 7.0 lacs   
4                                institutional sales   Rs. 6.8 lacs   

        Work_Experience                          Level Preferred_Location  \
0  6 Year(s) 3 Month(s)            Entry Level/Fresher              Delhi   
1  8 Year(s) 2 Month(s)            Entry Level/Fresher              Delhi   
2           22 Year(s)             Entry Level/Fresher              Delhi   
3            9 Year(s)             Entry Level/Fresher  Anywhere in India   
4  9 Year(s) 5 Month(s)  Junior - Executive, Assistant         Delhi/ NCR   

  Course(Highest_Education) Specialization(Highest_Education)  \
0                     B.Com                   Commerce Honors   
1                      B.A.                       Pass Course   
2                      12th                           Science   
3                     B.Com                   Commerce Honors   
4                  MBA/PGDM                         Marketing   

  Institute(Highest_Education) Course(2nd_Highest_Education)  \
0            Jiwaji University                           NaN   
1                       Others                           NaN   
2                       Others                          B.A.   
3                       Others                         B.Com   
4    Aligarh Muslim University                         B.Com   

  Specialization(2nd_Highest_Education) Institute(2nd_Highest_Education)  
0                                   NaN                              NaN  
1                                   NaN                              NaN  
2                          Anthropology                           Others  
3                       Commerce Honors                           Others  
4                         Commerce Pass                 Delhi University

In [19]:
# Display minimum and maximum dates of birth
dob_min: pd.Timestamp = df['DOB'].min()
dob_max: pd.Timestamp = df['DOB'].max()

print(f"Date of Birth Range:")
print(f"Minimum DOB: {dob_min}")
print(f"Maximum DOB: {dob_max}")
print(f"Date range spans: {dob_max - dob_min}")

Date of Birth Range:
Minimum DOB: 1900-01-01 00:00:00
Maximum DOB: 2008-09-03 00:00:00
Date range spans: 39692 days 00:00:00


In [18]:
# Get categorical columns excluding DOB
categorical_columns: list = df.select_dtypes(include=['object']).columns.tolist()
categorical_columns.remove('DOB') if 'DOB' in categorical_columns else None

# Display value counts for each categorical column
for column_name in categorical_columns:
    print(f"\n=== {column_name} ===")
    value_counts_result = df[column_name].value_counts()
    print(f"Cardinality: {len(value_counts_result)}")
    print(value_counts_result.head(10))  # Show top 10 values
    if len(value_counts_result) > 10:
        print(f"... and {len(value_counts_result) - 10} more values")
    print("-" * 50)


=== City ===
Cardinality: 13211
City
HYDERABAD        5641
BANGALORE        5503
RANGA REDDY      3893
PUNE             3752
VISAKHAPATNAM    2869
NEW DELHI        2353
MUMBAI           2299
PATNA            2203
KOLKATA          2144
SURAT            2143
Name: count, dtype: int64
... and 13201 more values
--------------------------------------------------

=== Functional_Area ===
Cardinality: 5405
Functional_Area
IT/Telecom - Software                              112088
Fresh Graduate-No Functional Area exp.              70060
Accounting/Tax/Company Secretary/Audit              66041
Sales/Business Development                          47189
Others                                              43302
Banks/Insurance/Financial Services                  41997
IT/Telecom - Hardware                               37228
Operations/Customer Service/Telecalling/Backend     36065
Engg. project Mgmt./site Engg./Civil Engg.          33597
Marketing/MR/Media Planning                         32936


## Dataset Description

| Feature | Datatype | Description | Cardinality |
|---------|----------|-------------|-------------|
| City | object | City where the job seeker is located | 13211 |
| DOB | datetime64[ns] | Date of birth of the job seeker | Continuous |
| Gender | object | Gender of the job seeker | 0 |
| Functional_Area | object | Functional area of work experience | 5405 |
| Area_of_Specialization | object | Area of specialization in work | 29547 |
| Industry | object | Industry of work experience | 7020 |
| Resume_Title | object | Title of the resume | 426939 |
| Key_Skills | object | Key skills of the job seeker | 448048 |
| Salary | object | Salary of the job seeker (in Rs or other currencies) | 6489 |
| Work_Experience | object | Work experience of the job seeker | 1101 |
| Level | object | Job level/position of the job seeker | 8 |
| Preferred_Location | object | Preferred work location of the job seeker | 24174 |
| Course(Highest_Education) | object | Highest educational qualification course | 37 |
| Specialization(Highest_Education) | object | Specialization in highest education | 172 |
| Institute(Highest_Education) | object | Institute where highest education was completed | 192 |
| Course(2nd_Highest_Education) | object | Second highest educational qualification course | 37 |
| Specialization(2nd_Highest_Education) | object | Specialization in second highest education | 168 |
| Institute(2nd_Highest_Education) | object | Institute where second highest education was completed | 156 |

## Data Quality Assessment and Preprocessing Strategy

### Critical Data Issues Identified

**Missing Data:**
- Gender column contains 0 non-null values and should be removed from analysis
- Date of birth range (1900-2008) appears anomalous and requires validation

**Data Quality Problems:**
- Salary stored as text with inconsistent formats (e.g., "Rs. X lacs")
- Work experience formatting inconsistencies (Years vs Year(s))
- High cardinality features requiring substantial preprocessing

### Feature Cardinality Analysis

**Extremely High Cardinality Features (>20k unique values):**
- Key_Skills: 448,048 unique values - requires skill extraction and categorization
- Resume_Title: 426,939 unique values - needs job title classification
- Area_of_Specialization: 29,547 unique values - requires domain grouping
- Preferred_Location: 24,174 unique values - needs geographic consolidation

**Moderate Cardinality Features (100-20k unique values):**
- City: 13,211 unique values - requires focus on major metropolitan areas
- Institute(Highest_Education): 192 unique values - manageable with tiering
- Institute(2nd_Highest_Education): 156 unique values - manageable

**Low Cardinality Features (<100 unique values):**
- Level: 8 unique values - suitable for ordinal encoding
- Course(Highest_Education): 37 unique values - appropriate for categorical analysis
- Course(2nd_Highest_Education): 37 unique values - appropriate for categorical analysis

### Preprocessing Strategy

**High Cardinality Feature Processing:**
- Key_Skills: Implement skill extraction algorithms, create standardized skill categories, consider TF-IDF vectorization
- Resume_Title: Develop job title classification system (technical, management, sales, etc.), evaluate feature importance
- Area_of_Specialization: Create domain-based groupings (IT, Finance, Marketing, Operations)
- Preferred_Location: Implement city tier classification, focus on major metropolitan areas

**Moderate Cardinality Feature Processing:**
- City: Retain top 20-30 cities by frequency, aggregate remaining as "Other Cities"
- Institute names: Create institutional tiers (IITs, NITs, State Universities, Private Institutions)

**Low Cardinality Feature Processing:**
- Level: Implement ordinal encoding based on hierarchical job levels
- Course features: Create education level categories and field classifications

**Special Feature Processing:**
- Salary: Extract numeric values, standardize currency formats, create salary brackets
- Work_Experience: Standardize to years, handle "Fresher" cases, create experience brackets
- DOB: Calculate age, validate date ranges, create age demographic brackets

### Feature Engineering Opportunities

**Interaction Features:**
- Education level × work experience interactions
- Salary per year of experience ratio
- Skills count and diversity metrics
- Geographic region × industry interactions

**Derived Features:**
- Age brackets and career stage classifications
- Education-to-experience ratio
- Skill specialization indices

### Encoding Strategy

**Categorical Encoding:**
- One-hot encoding for Level, Course categories, Institute tiers
- Ordinal encoding for education levels and experience brackets
- Target encoding for high cardinality features
- Text embeddings for skill and title features

This comprehensive preprocessing approach will transform the dataset into a machine learning-ready format while preserving essential information and reducing dimensionality.